data/lua/diversion.lua

```diff
@@ -88,20 +86,16 @@ local status_anim_update = function(is_undo)
                     }
                 end
         end
-        if changed_something and not is_undo then
-                wesnoth.wml_actions.on_undo {
-                        wml.tag.on_undo_diversion {
-                        }
-                }
+        if not is_undo then
+                wesnoth.game_events.set_undoable(true)
+                if changed_something then
+                        wesnoth.game_events.add_undo_actions(function(_)
+                                status_anim_update(true)
```

Comment:
In what order does the undo happen? From this code, it seems the units are already in position before whatever triggered this, so step 5 comes before step 6:

1. player moves unit with `diversion`
2. position of units update on map
3. moveto event triggers `status_anim_update` to recalculate the effects
4. player presses "undo"
5. position of units update on map
6. undo event triggers `status_anim_update` to recalculate the effects

However, consider an event that removes a unit from the map, for example a scenario where the player must move units to an exit point, and then they're immediately moved to the player's recall list. How can that be undone, unless the code for the undo runs and puts the unit back on the map before the movement is reversed? So in that case, step 6 must come before step 5.

Reply:
I think the code itself would just need to be written to handle it somehow. Like, in your example, it stores the ID and uses it to look up the unit on the recall list and put it back on the map.

Reply:
I think in theory it makes sense if the order of execution when undoing is the reverse of the order of execution during the original move (which would mean the when u doing first the move to event is undone and then the actual move). However I'm unsure how feasible this is, especially when there are also events like enter_hex and exit_hex that happen during the move.

Reply:
That does kinda make sense. It also breaks this diversion implementation. How would diversion work if the undo actions were triggered before the units were moved back to their prior configuration?

Reply:
I don't really know much about the animation code, so i am not sure what the code does currently, but the general idea is is to remember the previous state/changes and pass it to the undo handler.